client/dhcp_client.ml

```diff
@@ -23,36 +31,37 @@ let default_requests =
 
 let pp fmt p =
   let pr = Dhcp_wire.pkt_to_string in
-  match p with
-  | Selecting pkt -> Format.fprintf fmt "SELECTING.  Generated %s" @@ pr pkt
-  | Requesting (received, sent) -> Format.fprintf fmt
-      "REQUESTING. Received %s, and generated response %s" (pr received) (pr sent)
-  | Bound pkt -> Format.fprintf fmt "BOUND.  Received %s" @@ pr pkt
+  let pp_state fmt = function
+    | Selecting pkt -> Format.fprintf fmt "SELECTING.  Generated %s" @@ pr pkt
+    | Requesting (received, sent) -> Format.fprintf fmt
+        "REQUESTING. Received %s, and generated response %s" (pr received) (pr sent)
+    | Bound pkt -> Format.fprintf fmt "BOUND.  Received %s" @@ pr pkt
+    | Renewing (ack, request) -> Format.fprintf fmt
+        "RENEWING.  Have lease %s, generated request %s" (pr ack) (pr request)
+  in
+  Format.fprintf fmt "%s: %a" (Macaddr.to_string p.srcmac) pp_state p.state
 
-let lease = function
-  | Bound dhcpack -> Some dhcpack
+let lease {state; _} = match state with
+  | Bound dhcpack | Renewing (dhcpack, _) -> Some dhcpack
   | Requesting _ | Selecting _ -> None
 
-let offer ~dhcpdiscover ~dhcpoffer () =
+let xid {state; _} =
   let open Dhcp_wire in
-  (* TODO: make sure the offer contains everything we expect before we accept it *)
-  let options = [
-    Message_type DHCPREQUEST;
-    Request_ip dhcpoffer.yiaddr;
-    Server_identifier dhcpoffer.siaddr;
-  ] in
-  let options =
-    match find_parameter_requests dhcpdiscover.options with
-    | None -> options (* if this is the case, the user explicitly requested it; honor that *)
-    | Some p -> (Parameter_requests p) :: options
-  in
-  let dhcprequest = Constants.({
+  match state with
+  | Selecting p -> p.xid
+  | Requesting (_i, o) -> o.xid
+  | Bound a -> a.xid
+  | Renewing (_i, o) -> o.xid
```

Comment:
Thanks for calling this out.  I included the `xid` in `t` to make things simpler for an experimental UNIX client that isn't included in this PR, but the changes are probably not sensible outside of that context (and should be implemented differently in that case, very likely).  I'll look at this more closely next week and try to make it more consistent.